### PR TITLE
Updating trim on posted data to be compatible with CakePHP 3.7+

### LINKED ISSUE #206: CakePHP 3.6/4.0 way for trimming inputs?
In CakePHP 3.6 user `request->data = ...` is deprecated and going away in 4.0.  I like your solution for trimming all inputs at the request level as shown in your CommonComponent:

https://github.com/dereuromark/cakephp-tools/blob/8b8e902d6781cb3685a79685fc415f9fdb038c4a/src/Controller/Component/CommonComponent.php#L30

Do you have any suggestions for a version 3.6/4.0 friendly way of doing the same sort of thing?

Some others seem concerned that cleaning the request data is [a bad thing](https://discourse.cakephp.org/t/cakephp-3-4-how-to-change-request-data-programmaticly/2120) but I personally don't think it is that big of a deal.  Still we have to play nicely with the newer versions of the framework...

Any thoughts or suggestions would be appreciated.

### PR DESCRIPTION
**Pull request by challgren**
Fixes #206

### CODE REVIEW

File: src/Controller/Component/CommonComponent.php [hunk 1]
```diff
@@ -25,8 +26,15 @@ class CommonComponent extends Component {
 	 */
 	public function startup(Event $event) {
 		// Data preparation
-		if ($this->Controller->request->getData() && !Configure::read('DataPreparation.notrim')) {
-			$this->Controller->request->data = Utility::trimDeep($this->Controller->request->getData());
+		if ($this->Controller->getRequest()->getData() && !Configure::read('DataPreparation.notrim')) {
+			$request = $this->Controller->getRequest();
+			$new_data = Hash::flatten(Utility::trimDeep($this->Controller->getRequest()->getData()));
```
Comment (dereuromark) [inline src/Controller/Component/CommonComponent.php]: newData

File: src/Controller/Component/CommonComponent.php [hunk 2]
```diff
@@ -25,8 +26,15 @@ class CommonComponent extends Component {
...
+			foreach ($new_data as $k => $v) {
+				if ($this->Controller->getRequest()->getData($k) != $v) {
```
Comment (dereuromark) [inline src/Controller/Component/CommonComponent.php]: always strict check IMO

File: src/Controller/Component/CommonComponent.php (hunk 1 above)
Comment (dereuromark) [reply src/Controller/Component/CommonComponent.php]: whats the hash flatten here?

Comment (challgren) [reply src/Controller/Component/CommonComponent.php]: - [x] newData
- [x] HashFlatten: Was over thinking things its been removed

File: src/Controller/Component/CommonComponent.php (hunk 2 above)
Comment (challgren) [reply src/Controller/Component/CommonComponent.php]: - [x] Strict Check completed

**Comment by dereuromark** [comment]
Looks good.

File: src/Controller/Component/CommonComponent.php [hunk 3]
```diff
@@ -25,8 +25,15 @@ class CommonComponent extends Component {
 	 */
 	public function startup(Event $event) {
 		// Data preparation
-		if ($this->Controller->request->getData() && !Configure::read('DataPreparation.notrim')) {
-			$this->Controller->request->data = Utility::trimDeep($this->Controller->request->getData());
+		if ($this->Controller->getRequest()->getData() && !Configure::read('DataPreparation.notrim')) {
+			$request = $this->Controller->getRequest();
+			$newData = Utility::trimDeep($this->Controller->getRequest()->getData());
+			foreach ($newData as $k => $v) {
+				if ($this->Controller->getRequest()->getData($k) !== $v) {
```
Comment (dereuromark) [inline src/Controller/Component/CommonComponent.php]: should we use the same comparison on query and pass params too?

**Pull request merged**